src/tusb.c

```diff
@@ -452,6 +452,22 @@ void tu_print_mem(void const* buf, uint32_t count, uint8_t indent) {
   dump_str_line(buf8 - nback, nback);
 }
 
+int tu_printf_r(const char *format, ...)
+{
+  // Simple re-entrancy lock, not safe for multi-core MCU
+  static volatile bool lock = false;
```

Comment:
Even on a single core, `volatile` is not sufficient to prevent inconsistencies due to ISR execution interrupting another instance executing from user space. This needs an actual critical section, or similar.

Reply:
Could you make a real example with disassembly ?

Reply:
The problem is with this sequence:
```
  if (lock) {
    return -1;
  }
  // <--- ISR COMES HERE
  lock = true;
```
To make that work on uniprocessor with interrupt, then interrupt must be disabled between the test and the setting to true:

something like:

```
  disable_interrupts();
  if (lock) {
    return -1;
  }
  lock = true;
  enable_interrupts()  
```

Reply:
And this is of course wrong, more like:

```
  bool is_locked;
  disable_interrupts();
  is_locked = lock;
  if (!is_locked)
      lock = true;
  enable_interrupts()  
  if (is_locked)
      return (-1);
```

Reply:
Hi,
> The problem is with this sequence:
```
  if (lock) {
    return -1;
  }
  // <--- ISR COMES HERE
  lock = true;
```
When ISR happened here, `vprintf` hasn't been executed, so ISR will execute `vprintf` in it's context then return to normal context. There is no case where the execution of `vprintf` is interrupted.

Reply:
But: what protect your IRQ from interrupting another vprintf() call, outside tinyusb?

Reply:
BTW: I'm using a stm32h747 with tinyusb, and I would love to see your DWC2 DMA patch going in, which is why I was lurking....

Reply:
> But: what protect your IRQ from interrupting another vprintf() call, outside tinyusb?

Good question, currently `TU_LOG_*` are available for user space logging.

> BTW: I'm using a stm32h747 with tinyusb, and I would love to see your DWC2 DMA patch going in

We are few maintaining the repo, if you don't want to wait you can try the [testing](https://github.com/hathach/tinyusb/tree/testing) branch.